crates/shared/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -87,166 +97,365 @@ impl TryFrom<PoolData> for PoolInfo {
+                                tick.liquidity_net += BigInt::from(burn.amount);
+                            }
+                        }
+                    }
+                }
+                UniswapV3Event::Mint(mint) => {
+                    let tick_lower = BigInt::from(mint.tick_lower);
+                    let tick_upper = BigInt::from(mint.tick_upper);
+
+                    //liquidity tracks the liquidity on recent tick,
+                    // only need to update it if the new position includes the recent tick.
+                    if tick_lower <= pool.tick && pool.tick < tick_upper {
+                        pool.liquidity += mint.amount.into();
+                    }
+
+                    if let Some(ticks) = &mut pool.ticks {
+                        //todo optimize to map
+                        if ticks.iter().all(|tick| tick.tick_idx != tick_lower) {
+                            ticks.push(TickData {
+                                id: address.to_string() + "#" + &tick_lower.to_string(),
+                                tick_idx: tick_lower.clone(),
+                                liquidity_net: 0.into(),
+                                pool_address: address,
+                            });
+                        }
+
+                        if ticks.iter().all(|tick| tick.tick_idx != tick_upper) {
+                            ticks.push(TickData {
+                                id: address.to_string() + "#" + &tick_upper.to_string(),
+                                tick_idx: tick_upper.clone(),
+                                liquidity_net: 0.into(),
+                                pool_address: address,
+                            });
+                        }
+
+                        for tick in ticks {
+                            if tick.tick_idx == tick_lower {
+                                tick.liquidity_net += BigInt::from(mint.amount);
+                            }
+                            if tick.tick_idx == tick_upper {
+                                tick.liquidity_net -= BigInt::from(mint.amount);
+                            }
+                        }
+                    }
+                }
+                UniswapV3Event::Swap(swap) => {
+                    pool.tick = BigInt::from(swap.tick);
+                    pool.liquidity = swap.liquidity.into();
+                    pool.sqrt_price = swap.sqrt_price_x96;
+                }
             }
         }
+    }
+}
 
-        tracing::debug!("periodic update ended");
-        Ok(())
+#[async_trait::async_trait]
+impl Maintaining for UniswapV3PoolFetcher {
+    async fn run_maintenance(&self) -> Result<()> {
+        let (result1, result2) = futures::join!(
+            self.events.run_maintenance(),
+            self.checkpoint.update_missing_pools()
+        );
+        ensure!(result1.is_ok(), result1.unwrap_err());
+        ensure!(result2.is_ok(), result2.unwrap_err());
```

Comment:
Or if you don't like try_join
```suggestion
    result1?;
    result2?;
```

Reply:
Taken. Thanks